iib/common/tracing.py

```diff
@@ -0,0 +1,135 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+"""Configures the Global Tracer Provider and exports the traces to the OpenTelemetry Collector.
+
+The OpenTelemetry Collector is configured to receive traces via OTLP over HTTP.
+The OTLP exporter is configured to use the environment variables defined in the ansible playbook.
+
+Usage:
+    @instrument_tracing()
+    def func():
+        pass
+
+"""
+import os
+import functools
+import logging
+from typing import Dict
+from opentelemetry import trace
+from opentelemetry.trace import SpanKind, Status, StatusCode
+from opentelemetry.sdk.resources import Resource, SERVICE_NAME
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import (
+    BatchSpanProcessor,
+)
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.trace.propagation import (
+    set_span_in_context,
+)
+from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
+from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+
+
+log = logging.getLogger(__name__)
+propagator = TraceContextTextMapPropagator()
+
+
+class TracingWrapper:
+    """Wrapper class that will wrap all methods of calls with the instrument_tracing decorator."""
+
+    __instance = None
+
+    def __new__(cls):
+        """Create a new instance if one does not exist."""
+        if TracingWrapper.__instance is None:
+            log.info('Creating TracingWrapper instance')
+            cls.__instance = super().__new__(cls)
+            otlp_exporter = OTLPSpanExporter(
+                endpoint=f"{os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT')}/v1/traces",
+            )
+            cls.provider = TracerProvider(
+                resource=Resource.create({SERVICE_NAME: os.getenv('OTEL_SERVICE_NAME') or 'iib'})
+            )
+            cls.processor = BatchSpanProcessor(otlp_exporter)
+            cls.provider.add_span_processor(cls.processor)
+            trace.set_tracer_provider(cls.provider)
+            set_global_textmap(propagator)
+            cls.tracer = trace.get_tracer(__name__)
+        return cls.__instance
+
+
+def instrument_tracing(
+    func=None,
+    span_name: str = '',
+    attributes: Dict = {},
+):
+    """
+    Instrument tracing for a function.
+
+    :param func: The function to be decorated.
+    :param span_name: The name of the span to be created.
+    :param attributes: The attributes to be added to the span.
```

Comment:
@raks-tt thanks a lot for opening this PR, I'm using it as an example for another project.

By observing this decorator I noticed the parameter `attributes` isn't being used anywhere. Is this correct?

Reply:
@JAVGan - Yes currently it is not.